packages/text-editor/src/components/CollaborativeTextEditor.svelte

```diff
@@ -0,0 +1,393 @@
+  export let boundary: HTMLElement | undefined = undefined
+
+  export let attachFile: FileAttachFunction | undefined = undefined
+  export let canShowPopups = true
+
+  const dispatch = createEventDispatcher()
+
+  const token = getMetadata(presentation.metadata.Token) ?? ''
+  const collaboratorURL = getMetadata(textEditorPlugin.metadata.CollaboratorUrl) ?? ''
+
+  const ydoc = getContext<YDoc>(CollaborationIds.Doc) ?? new YDoc()
+  const contextProvider = getContext<TiptapCollabProvider>(CollaborationIds.Provider)
+
+  const provider: TiptapCollabProvider =
+    contextProvider ??
+    new TiptapCollabProvider({
+      url: collaboratorURL,
+      name: documentId,
+      document: ydoc,
+      token,
+      parameters: {
+        initialContentId,
+        targetContentId
+      }
+    })
+
+  let loading = true
+  void provider.loaded.then(() => (loading = false))
+
+  const currentUser = getCurrentAccount()
+
+  let editor: Editor
+  let element: HTMLElement
+  let textToolbarElement: HTMLElement
+  let imageToolbarElement: HTMLElement
+
+  let placeHolderStr: string = ''
+
+  $: ph = translate(placeholder, {}, $themeStore.language).then((r) => {
+    placeHolderStr = r
+  })
+
+  $: dispatch('editor', editor)
+
+  const editorHandler: TextEditorHandler = {
+    insertText: (text) => {
+      editor?.commands.insertContent(text)
+    },
+    insertTemplate: (name, text) => {
+      editor?.commands.insertContent(text)
+    },
+    focus: () => {
+      focus()
+    }
+  }
+
+  function handleAction (a: RefAction, evt?: Event): void {
+    a.action(evt?.target as HTMLElement, editorHandler)
+  }
+
+  $: commandHandler = textEditorCommandHandler(editor)
+
+  export function commands (): TextEditorCommandHandler | undefined {
+    return commandHandler
+  }
+
+  export function takeSnapshot (snapshotId: string): void {
+    copyDocumentContent(documentId, snapshotId, { provider }, initialContentId)
+  }
+
+  export function copyField (srcFieldId: string, dstFieldId: string): void {
+    copyDocumentField(documentId, srcFieldId, dstFieldId, { provider }, initialContentId)
+  }
+
+  export function isEditable (): boolean {
+    return editor?.isEditable ?? false
+  }
+
+  let needFocus = false
+  let focused = false
+  let posFocus: FocusPosition | undefined = undefined
+
+  export function focus (position?: FocusPosition): void {
+    posFocus = position
+    needFocus = true
+  }
+
+  export function isFocused (): boolean {
+    return focused
+  }
+
+  $: if (editor !== undefined && needFocus) {
+    if (!focused) {
+      editor.commands.focus(posFocus)
+      posFocus = undefined
+    }
+    needFocus = false
+  }
+
+  function handleFocus (): void {
+    needFocus = true
+  }
+
+  $: if (editor !== undefined) {
+    editor.setEditable(!readonly)
+  }
+
+  $: showTextStyleToolbar =
+    (!readonly || textFormatCategories.length > 0 || textNodeActions.length > 0) && canShowPopups
+
+  $: tippyOptions = {
```

Comment:
As I remember you wanted to discard tippy?

Reply:
yes, it will be discarded